pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentSubscription.java

```diff
@@ -76,13 +90,38 @@
     // for connected subscriptions, message expiry will be checked if the backlog is greater than this threshold
     private static final int MINIMUM_BACKLOG_FOR_EXPIRY_CHECK = 1000;
 
+    // Map to keep track of message ack by each txn.
+    private final ConcurrentOpenHashMap<TxnID, ConcurrentOpenHashSet<Position>> pendingAckMessagesMap;
+
+    // Messages acked by ongoing transaction, pending transaction commit to materialize the acks. For faster look up.
+    // Using hashset as a message should only be acked once by one transaction.
+    private final ConcurrentOpenHashSet<Position> pendingAckMessages;
```

Comment:
I think this has to be a reverse map from `Position` to `TxnID`, no? You need to use that for conflict detection, no?

Reply:
I actually used another ConcurrentOpenHashSet to keep all the unacked messages(Position) and it's just for conflict detection. 
I'm not sure if there'll be performance/space benefit for using one reverse map vs a map + a set. Or maybe you think using one reverse map is cleaner?

Reply:
I think using one reverse map would be cleaner.

Reply:
I tried to use one reverse map of Map<Position, TxnId>, it is cleaner when check if position is acked by ongoing transaction.
But when we try to commit or abort a transaction given a TxnId it's not very easy as we need to first  traverse by value and get all positions for that transaction, materialize to ledger,  then traverse again to remove all entry for a given TxnId. I'm worry about performance for the extra traverse we need. (I'm using ConcurrentOpenHashMap which doesn't have access to entry set.)
I'll keep it as it for this pr as it's only in memory ack, if we need to make change here I can do it in subsequent pr.